### ISSUE TITLE
windows spawning on the wrong screen/moving to another screen after layout change

### ISSUE THREAD

**Issue by paul-axe** [awesomeWM/awesome#1685]
Output of `awesome --version`:

awesome --version
awesome v4.1-19-gb81b214-dirty (Technologic)
 • Compiled against Lua 5.3.1 (running with Lua 5.3)
 • D-Bus support: ✔
 • execinfo support: ✔
 • xcb-randr version: 1.4
 • LGI version: 0.9.1

**How to reproduce the issue:**

I have two monitors configuration with default rc.lua. When i trying to open window on the right monitor, it appears on the first screen (not always). If window is already on the second monitor, it moves to first one when i changing layout to tile.  

I found that the issue is in `client_resize` function. `display_area_get` returns geometry for first screen only (not sure if it should return all screens geometry together, or only current screen). I patched it to get current screen geometry, so it works for me now. You can find patch here: [https://pastebin.com/dyRYkngQ](https://pastebin.com/dyRYkngQ)

**Comment by Veratil** [comment]
All tests run successfully for me after applying the patch. I'll test how this patch works next time I'm in my office with multiple screens where I know I can reproduce the issue.

**Comment by blueyed** [comment]
Thanks for the patch!
Can you create a PR with it, please?

**Comment by psychon** [comment]
> I found that the issue is in client_resize function. display_area_get returns geometry for first screen only (not sure if it should return all screens geometry together, or only current screen). 

It should return the size of the root window. The root window should be large enough to contain all RANDR outputs. If the size of the root window is wrong, then setting the wallpaper should not work. How do you start awesome? Is it concurrently with RandR changes? Do you set the wallpaper through awesome? If not, could you try if that works correctly or if it cuts of something and only works on the first screen? `awesome-client 'require("gears.wallpaper").set("#ff0000")` should, for example, set everything to a red background (but in your case I would expect some lack border)

What output does `xdpyinfo | grep dimensions` give you? What does `xrandr` print? What does `awesome-client 'return root.size()'` print?

**Comment by paul-axe** [comment]
Then my patch is incorrect :) I use only clients screen geometry.

I start awesome using GDM with default session from ubuntu (or with Xephyr for tests, but i have this issue for both cases). I use deafult awesome config for tests, and it sets wallpapers to both screens. Setting wallpaper to red works too. 

```
$ xdpyinfo | grep dimensions
  dimensions:    1600x600 pixels (542x203 millimeters)
$ awesome-client 'return root.size()'
   double 800
   double 600
```

**Comment by psychon** [comment]
Weird. The values in `globalconf.screen->{width,height}_in_pixels` come from the X11 server when awesome connects to the server. So I guess something makes a RandR screen change after awesome is connecting. However, in that case we should get events telling us about this that are handled here:
https://github.com/awesomeWM/awesome/blob/39dca5220c3aafc5d8a82743d97c5be82ad3eff5/event.c#L446-L450
https://github.com/awesomeWM/awesome/blob/39dca5220c3aafc5d8a82743d97c5be82ad3eff5/event.c#L816-L823
Somehow you manage to "get around" that.

> I use deafult awesome config for tests, and it sets wallpapers to both screens. Setting wallpaper to red works too.
> [...]
> $ awesome-client 'return root.size()'
>   double 800
>   double 600

Weird. `gears.wallpaper` uses `root.size()` to figure out how large the wallpaper should be. Thus, your second monitor should not have a wallpaper. What is going on here?!?
https://github.com/awesomeWM/awesome/blob/master/lib/gears/wallpaper.lua#L37

**Comment by paul-axe** [comment]
I thinkg it works because actually it uses geometry of current screen and called for every screen.

```
    local geom = s and s.geometry or root_geometry()
    ...
    cr:translate(geom.x, geom.y)
    cr:rectangle(0, 0, geom.width, geom.height)
    cr:clip()
```

**Comment by paul-axe** [comment]
I'll just figured out that issue appears only when i'm running awesome from gdm session or xephyr. I'll test it a little bit later.

**Comment by paul-axe** [comment]
Well, now i can reproduce this issue only in Xephyr. Weird.

**Comment by Veratil** [comment]
For more details, I can always reproduce the issue with a screen layout like this:
```
+---------+----+
|         | 4  |
+----+----+----+
| 3  | 2  | 1  |
+----+----+----+
```
When I'm on screen 1, spawning anything will pop up on screen 4.

xrandr info (pruned modes):
```
$ xrandr
Screen 0: minimum 320 x 200, current 5520 x 2104, maximum 16384 x 16384
LVDS-1 connected primary 1920x1080+3600+1024 (normal left inverted right x axis y axis) 344mm x 193mm
VGA-1 connected 1280x1024+3600+0 (normal left inverted right x axis y axis) 376mm x 301mm
DP-1 disconnected (normal left inverted right x axis y axis)
DP-2 connected 1920x1080+1680+1024 (normal left inverted right x axis y axis) 509mm x 286mm
DP-3 connected 1680x1050+0+1024 (normal left inverted right x axis y axis) 474mm x 296mm
```

**Comment by Elv13** [comment]
@psychon Was this issue improved/fixed by #1780? (also: https://github.com/awesomeWM/awesome/pull/1687)

**Comment by psychon** [comment]
> I found that the issue is in client_resize function. display_area_get returns geometry for first screen only

Fixed by c57208d1a820a9e11085844a00204e37a35a54cd (which removed this code).
There is still some issue on that we somehow ended up in a situation where awesome didn't know the root window size. The only other effect of that that I can think of right now would be "wallpaper is not set correctly", but since this was not reported here, I'll stay confused and will just close this.

**Comment by Veratil** [comment]
I'll test next time I'm at work to see if this fixed my case.
